### PR TITLE
Merge `develop` into `next`

### PR DESCRIPTION
**Pull request by jathanism**
# Closes: #DNE
# What's Changed
Request to merge `develop` into `next` to get latest downstream changes into the v2.0.0 train.

# TODO

- [ ] Explanation of Change(s)
- [ ] Added change log fragment(s) (for more information see [the documentation](https://docs.nautobot.com/projects/core/en/stable/development/#creating-changelog-fragments))
- [ ] Attached Screenshots, Payload Example
- [ ] Unit, Integration Tests
- [ ] Documentation Updates (when adding/changing features)
- [ ] Example Plugin Updates (when adding/changing features)
- [ ] Outline Remaining Work, Constraints from Design

### CODE REVIEW

File: nautobot/dcim/models/devices.py [hunk 1]
```diff
@@ -708,6 +695,50 @@ def save(self, *args, **kwargs):
...
+            self.rack.name if self.rack else None,
+            self.position,
+            self.get_face_display(),
+            self.device_redundancy_group.slug if self.device_redundancy_group else None,
+            self.device_redundancy_group_priority,
+            self.secrets_group.name if self.secrets_group else None,
+            self.primary_ip if self.primary_ip else None,
+            self.comments,
+        )
+
```
Comment (glennmatthews) [inline nautobot/dcim/models/devices.py:741]: remove - `to_csv` is not needed in 2.x.

File: nautobot/docs/models/extras/computedfield.md [hunk 2]
```diff
@@ -41,6 +41,9 @@ Computed field templates can utilize the context of the object the field is bein
 {{ obj.site.name }}
 ```
 
+!!! note
+    To access custom fields of an object within a template, use the `cf` attribute. For example, `{{ obj.cf.color }}` will return the value (if any) for a custom field named `color` on `obj`.
```
Comment (glennmatthews) [inline nautobot/docs/models/extras/computedfield.md:45]:
```suggestion
    To access custom fields of an object within a template, use the `cf` attribute. For example, `{{ obj.cf.color }}` will return the value (if any) for the custom field with a key of `color` on `obj`.
```

File: nautobot/docs/models/extras/customlink.md [hunk 3]
```diff
@@ -51,6 +51,9 @@ As another example, if you wanted to show only devices belonging to a certain ma
 
 The link will only appear when viewing a device with a manufacturer name of "Cisco."
 
+!!! note
+    To access custom fields of an object within a template, use the `cf` attribute. For example, `{{ obj.cf.color }}` will return the value (if any) for a custom field named `color` on `obj`.
```
Comment (glennmatthews) [inline nautobot/docs/models/extras/customlink.md:55]:
```suggestion
    To access custom fields of an object within a template, use the `cf` attribute. For example, `{{ obj.cf.color }}` will return the value (if any) for the custom field with a key of `color` on `obj`.
```

File: nautobot/docs/models/extras/exporttemplate.md [hunk 4]
```diff
@@ -16,7 +16,8 @@ Height: {{ rack.u_height }}U
 {% endfor %}
 ```
 
-To access custom fields of an object within a template, use the `cf` attribute. For example, `{{ obj.cf.color }}` will return the value (if any) for a custom field named `color` on `obj`.
+!!! note
+    To access custom fields of an object within a template, use the `cf` attribute. For example, `{{ obj.cf.color }}` will return the value (if any) for a custom field named `color` on `obj`.
```
Comment (glennmatthews) [inline nautobot/docs/models/extras/exporttemplate.md:20]:
```suggestion
    To access custom fields of an object within a template, use the `cf` attribute. For example, `{{ obj.cf.color }}` will return the value (if any) for the custom field with a key of `color` on `obj`.
```

File: nautobot/docs/models/extras/jobbutton.md [hunk 5]
```diff
@@ -73,6 +75,9 @@ The button will only appear if they have the permission to run jobs.
 
 ![Job Buttons on Site object](../../media/models/site_jobbuttons.png "Job Buttons on Site object")
 
+!!! note
+    To access custom fields of an object within a template, use the `cf` attribute. For example, `{{ obj.cf.color }}` will return the value (if any) for a custom field named `color` on `obj`.
```
Comment (glennmatthews) [inline nautobot/docs/models/extras/jobbutton.md:79]:
```suggestion
    To access custom fields of an object within a template, use the `cf` attribute. For example, `{{ obj.cf.color }}` will return the value (if any) for the custom field with a key of `color` on `obj`.
```

File: nautobot/docs/models/extras/secret.md [hunk 6]
```diff
@@ -29,3 +29,6 @@ In some cases you may have a collection of closely related secrets values that a
 
 - A "Device Password" secret could use the *Text File* provider and specify the file `path` as `"/opt/nautobot/device_passwords/{{ obj.site.slug }}/{{ obj.name }}.txt"`, so that a device `csr1` at site `nyc` would be able to retrieve its password value from `/opt/nautobot/device_passwords/nyc/csr1.txt`.
 - A "Git Token" secret could use the *Environment Variable* provider and specify the `variable` name as `"GIT_TOKEN_{{ obj.slug | replace('-', '_') | upper }}"`, so that a Git repository `golden-config` would be able to retrieve its token value from `$GIT_TOKEN_GOLDEN_CONFIG`.
+
+!!! note
+    To access custom fields of an object within a template, use the `cf` attribute. For example, `{{ obj.cf.color }}` will return the value (if any) for a custom field named `color` on `obj`.
```
Comment (glennmatthews) [inline nautobot/docs/models/extras/secret.md:34]:
```suggestion
    To access custom fields of an object within a template, use the `cf` attribute. For example, `{{ obj.cf.color }}` will return the value (if any) for the custom field with a key of `color` on `obj`.
```

File: nautobot/extras/models/jobs.py [hunk 7]
```diff
@@ -248,6 +248,12 @@ def class_path(self):
 
     @property
     def latest_result(self):
+        """
+        Return the most recent JobResult object associated with this Job.
+
+        Note that, as a performance optimization for this function's repeated use in
+        JobListview, the returned object only includes its `status` field.
```
Comment (glennmatthews) [inline nautobot/extras/models/jobs.py:255]: Is this accurate? I don't see `.only("status")` at line 252/258 below?

Comment (jathanism) [reply nautobot/extras/models/jobs.py:255]: Nope. Fixing. Bad merge choice.

File: nautobot/dcim/models/devices.py (hunk 1 above)
Comment (jathanism) [reply nautobot/dcim/models/devices.py:741]: It looks like `GitRepository.to_csv()` does still exist in `next`, however. Was this an oversight?

Comment (glennmatthews) [reply nautobot/dcim/models/devices.py]: I'd imagine so. Not sure how I missed that one. :)

File: nautobot/docs/release-notes/version-1.5.md [hunk 8]
```diff
@@ -148,6 +148,72 @@ A number of mixin classes have been renamed and/or relocated for improved self-c
 | `TenancyFilterSet`             | `TenancyModelFilterSetMixin`                 |
 
 <!-- towncrier release notes start -->
+## v1.5.21 (2023-06-12)
+
+### Added
+
+- [#3806](https://github.com/nautobot/nautobot/issues/3806) - Added instructions and examples for SAML SSO using Okta as the IdP.
+- [#3811](https://github.com/nautobot/nautobot/issues/3811) - Added a note that addresses UWSGI buffer size concerns with Azure SSO in `nautobot/docs/configuration/authentication/sso.md`.
+- [#3897](https://github.com/nautobot/nautobot/issues/3897) - Adds log message when a secrets group for a git repository doesn't yield a token.
```
Comment (glennmatthews) [inline nautobot/docs/release-notes/version-1.5.md:157]: I don't see the diffs from #3897 (`nautobot/extras/datasources/git.py`) in this PR.

File: nautobot/extras/signals.py [hunk 9]
```diff
@@ -21,7 +22,10 @@
 from .models import CustomField, DynamicGroup, DynamicGroupMembership, GitRepository, JobResult, ObjectChange
 from .webhooks import enqueue_webhooks
 
-logger = logging.getLogger(__name__)
+
+# thread safe change context state variable
+change_context_state = contextvars.ContextVar("change_context_state", default=None)
+logger = logging.getLogger("nautobot.extras.signals")
```
Comment (glennmatthews) [inline nautobot/extras/signals.py:28]: Should probably revert this:

```suggestion
logger = logging.getLogger(__name__)
```

File: nautobot/extras/views.py [hunk 10]
```diff
@@ -1463,7 +1463,7 @@ class JobResultListView(generic.ObjectListView):
     List JobResults
     """
 
-    queryset = JobResult.objects.select_related("job_model", "user").prefetch_related("logs")
+    queryset = JobResult.objects.defer("result").select_related("job_model", "user").prefetch_related("logs")
```
Comment (glennmatthews) [inline nautobot/extras/views.py:1466]: This and the change below are different than the changes made in 7ffed896ec5ca4f61d0c23a637306192f34df2fd, might be worth looking at very closely to see if they're causing the CI failures.

File: nautobot/extras/context_managers.py [hunk 11]
```diff
@@ -91,22 +90,15 @@ def change_logging(change_context):
```
Comment (glennmatthews) [inline nautobot/extras/context_managers.py:90]: Missing change above here to delete the `curry()` function as no longer used/necessary.

File: nautobot/docs/release-notes/version-1.5.md [hunk 12]
```diff
@@ -148,6 +148,72 @@ A number of mixin classes have been renamed and/or relocated for improved self-c
...
+- [#3549](https://github.com/nautobot/nautobot/issues/3549) - Updated `django` to `~3.2.19` to address `CVE-2023-31047`.
+- [#3549](https://github.com/nautobot/nautobot/issues/3549) - Updated `mkdocs` to `~1.4.3`.
+- [#3549](https://github.com/nautobot/nautobot/issues/3549) - Updated `psycopg2-binary` to `~2.9.6`.
+- [#3698](https://github.com/nautobot/nautobot/issues/3698) - Updated `social-auth-core` to `~4.4.0` to permit addressing `CVE-2022-2309`.
+- [#3753](https://github.com/nautobot/nautobot/issues/3753) - Updated indirect dev dependency `pymdown-extensions` to `10.0` to address `CVE-2023-32309`.
+
+### Fixed
+
+- [#3704](https://github.com/nautobot/nautobot/issues/3704) - Fixed GitRepository fetching on Home Page when getting repo-based Job's name.
+- [#3726](https://github.com/nautobot/nautobot/issues/3726) - Fixed a `KeyError` when filtering Cables in the UI by `termination_a_type` or `termination_b_type`.
```
Comment (glennmatthews) [inline nautobot/docs/release-notes/version-1.5.md:209]: I don't see the diffs from #3727 here - they were in `nautobot/utilities/utils.py` in develop but should go into `nautobot/core/utils/filtering.py` in `next`.

File: nautobot/dcim/models/devices.py (hunk 1 above)
Comment (jathanism) [reply nautobot/dcim/models/devices.py]: Cool I'm leaving it there for now until I can get tests passing.

**Review by glennmatthews — APPROVED** [review]

**Pull request merged**